### PR TITLE
Add session key types for new consensus implementation to Dolphin and Calamari (Nimbus Stage 0)

### LINKED ISSUE #668: Nimbus Consensus Migration Stage 0
This is phase 0 of the phased Nimbus consensus rollout described in #486 

Nimbus Consensus uses different validator keys than our currently used Aura

To allow collator runners to upgrade their nodes in preparation for the switch to Nimbus, we add the required key to the KMA and DOL runtimes.
For futureproofing, we also enable a VRF key type in addition to the Nimbus key
( see https://github.com/PureStake/moonbeam/pull/1480 )
This will allow us to change to a VRF lottery style block producer election without asking collators to recreate their keychain again.

In summary:
- [ ] Merge code with wrapped nimbus opaque key with OneSessionHandler and noop on_new_session #646 
- [ ] Issue runtime upgrade to v3.2.1
- [ ] publish community post to call set_keys in advance of nimbus enabling
- [ ] update keys of manta-run collators
  - [ ] dolphin
  - [ ] calamari

### PR DESCRIPTION
**Pull request by Garandor**
This PR does not contain any functional changes to KMA or DOL with the following two exceptions:
⚠️ The `rotateKeys` RPC call now returns 3 keys instead of 1
⚠️ The `setKeys` extrinsic now requires the above 3 keys as parameters instead of just the aura key
Nimbus consensus will be enabled in a later PR.

- [x] Refactor all (aura-)consensus specifics and dependencies out of manta-primitives into a new session-key-primitives crate inspired by Moonbeam's PR above
- [x] Extract the SessionKeys type from the runtimes into the common session-key-primitives module.
Attempted but failed as the necessary impl_opaque_keys! macro does not support generics

related to #668 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests.
- [x] Updated relevant documentation in the code.
- [x] Added **one** line describing your change in `<branch>/CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] If runtime changes, need to update the version numbers properly:
   * `authoring_version`: The version of the authorship interface. An authoring node will not attempt to author blocks unless this is equal to its native runtime.
   * `spec_version`: The version of the runtime specification. A full node will not attempt to use its native runtime in substitute for the on-chain Wasm runtime unless all of spec_name, spec_version, and authoring_version are the same between Wasm and native.
   * `impl_version`: The version of the implementation of the specification. Nodes are free to ignore this; it serves only as an indication that the code is different; as long as the other two versions are the same then while the actual code may be different, it is nonetheless required to do the same thing. Non-consensus-breaking optimizations are about the only changes that could be made which would result in only the impl_version changing.
   * `transaction_version`: The version of the extrinsics interface. This number must be updated in the following circumstances: extrinsic parameters (number, order, or types) have been changed; extrinsics or pallets have been removed; or the pallet order in the construct_runtime! macro or extrinsic order in a pallet has been changed. You can run the `metadata_diff.yml` workflow for help. If this number is updated, then the `spec_version` must also be updated 
- [ ] Verify benchmarks & weights have been updated for any modified runtime logics
- [ ] If importing a new pallet, choose a proper module index for it, and allow it in `BaseFilter`. Ensure **every** extrinsic works from front-end. If there's corresponding tool, ensure both work for each other.
- [ ] If needed, update our Javascript/Typescript APIs. These APIs are officially used by exchanges or community developers.
- [ ] If modifying existing runtime storage items, make sure to implement storage migrations for the runtime and test them with `try-runtime`. This includes migrations inherited from upstream changes, and you can search the diffs for modifications of `#[pallet::storage]` items to check for any.

### CODE REVIEW

File: session-keys/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+scale-info = { version = "2.0", default-features = false, features = ["derive"] }
+sp-application-crypto = { git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.18", default-features = false }
+sp-consensus-aura = { git = 'https://github.com/paritytech/substrate.git', default-features = false, branch = "polkadot-v0.9.18" }
+sp-core = { git = 'https://github.com/paritytech/substrate.git', default-features = false, branch = "polkadot-v0.9.18" }
+sp-runtime = { git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.18", default-features = false }
+
+[features]
+default = ["std"]
+std = [
+  "nimbus-primitives/std",
```
Comment (ghzlatarev) [inline session-keys/Cargo.toml]: shouldn't manta-primitives also be included

File: session-keys/src/nimbus.rs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+// Manta is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Manta.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Helpers for handling nimbus keys
+
```
Comment (ghzlatarev) [inline session-keys/src/nimbus.rs]: Shouldn't we put #![cfg_attr(not(feature = "std"), no_std)] in all files that go into the runtime

File: node/src/chain_specs/dolphin.rs [hunk 3]
```diff
@@ -38,9 +40,14 @@ pub type DolphinChainSpec = sc_service::GenericChainSpec<GenesisConfig, Extensio
 
 /// Generate the dolphin session keys from individual elements.
 ///
-/// The input must be a tuple of individual keys (a single arg for now since we have just one key).
-pub fn dolphin_session_keys(keys: AuraId) -> dolphin_runtime::opaque::SessionKeys {
-    dolphin_runtime::opaque::SessionKeys { aura: keys }
+/// The input must be a tuple of individual keys (a single arg for now sinc we have just one key).
```
Comment (ghzlatarev) [inline node/src/chain_specs/dolphin.rs]: nit: `sinc` spelling

File: runtime/calamari/src/lib.rs [hunk 4]
```diff
@@ -752,8 +780,50 @@ pub type Executive = frame_executive::Executive<
...
+
+        for it in pre_migration_keys
+            .iter()
+            .zip(new_owners_and_aura_keys.iter())
+        {
+            let ((old_owner, old_key), (new_owner, new_key)) = iter;
+            ensure!(old_owner == new_owner, "owner changed");
+            ensure!(old_key.aura == new_key.aura, "key changed");
+            // ensure!(old_key.aura == new_key.nimbus ,"key changed");
+            // ensure!(old_key.aura == new_key.vrf ,"key changed");
```
Comment (ghzlatarev) [inline runtime/calamari/src/lib.rs]: commented code.

File: runtime/calamari/src/lib.rs [hunk 5]
```diff
@@ -752,8 +780,50 @@ pub type Executive = frame_executive::Executive<
     frame_system::ChainContext<Runtime>,
     Runtime,
     AllPalletsReversedWithSystemFirst,
-    (),
+    UpgradeSessionKeys,
```
Comment (ghzlatarev) [inline runtime/calamari/src/lib.rs:778]: Please run the try-runtime GHA

File: session-keys/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+pub trait KeysLookup<AuthorId, Keys> {
+    fn lookup_keys(author: &AuthorId) -> Option<Keys>;
+}
+
+// A dummy impl used in simple tests
+impl<AuthorId, Keys> KeysLookup<AuthorId, Keys> for () {
+    fn lookup_keys(_: &AuthorId) -> Option<Keys> {
+        None
+    }
+}
```
Comment (ghzlatarev) [inline session-keys/src/lib.rs]: Not sure why this is separate from `primitives` ?

File: runtime/calamari/src/nimbus_session_adapter.rs [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+//
+// Manta is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Manta.  If not, see <http://www.gnu.org/licenses/>.
+
+// TODO: This file should be part of primitives along with session-keys
```
Comment (ghzlatarev) [inline runtime/calamari/src/nimbus_session_adapter.rs]: Forgotten todo

File: runtime/dolphin/src/nimbus_session_adapter.rs [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+/// This adapts pallet AuthorInherent to be compatible with pallet session
+/// making it suitable as a SessionKey entry
+pub struct AuthorInherentWithNoOpSession<T: Config>(pub AuthorInherent<T>);
+
+impl<T: Config> BoundToRuntimeAppPublic for AuthorInherentWithNoOpSession<T> {
+    type Public = <AuthorInherent<T> as BoundToRuntimeAppPublic>::Public;
+}
+
+impl<T: Config> OneSessionHandler<T::AccountId> for AuthorInherentWithNoOpSession<T> {
```
Comment (ghzlatarev) [inline runtime/dolphin/src/nimbus_session_adapter.rs:35]: I guess these empty implementations are just to get it to compile then ?

File: runtime/dolphin/src/lib.rs [hunk 9]
```diff
@@ -740,7 +768,50 @@ pub type Executive = frame_executive::Executive<
...
+
+        for it in pre_migration_keys
+            .iter()
+            .zip(new_owners_and_aura_keys.iter())
+        {
+            let ((old_owner, old_key), (new_owner, new_key)) = iter;
+            ensure!(old_owner == new_owner, "owner changed");
+            ensure!(old_key.aura == new_key.aura, "key changed");
+            // ensure!(old_key.aura == new_key.nimbus ,"key changed");
+            // ensure!(old_key.aura == new_key.vrf ,"key changed");
```
Comment (ghzlatarev) [inline runtime/dolphin/src/lib.rs]: forgotten commented code

File: runtime/calamari/src/lib.rs (hunk 5 above)
Comment (Garandor) [reply runtime/calamari/src/lib.rs:778]: Completed here: https://github.com/Manta-Network/Manta/actions/runs/2634849527

File: session-keys/Cargo.toml (hunk 1 above)
Comment (Garandor) [reply session-keys/Cargo.toml]: good catch, thanks!

File: session-keys/src/nimbus.rs (hunk 2 above)
Comment (Garandor) [reply session-keys/src/nimbus.rs]: good catch, thanks!

File: session-keys/src/lib.rs (hunk 6 above)
Comment (Garandor) [reply session-keys/src/lib.rs]: Main benefit: `manta-primitives` is now agnostic to all consensus crates,  and has only two kinds of deps
- low-level substrate 
- manta-controlled crates

It's mainly separation of concerns. Anything consensus-related is bundled together
=> pallets that don't care about `ValidatorIds` and other consensus specifics don't need to know what an AuraId is.
=> `manta-primitives` will not change no matter what other changes to consensus-related stuff i make. Also consensus crate content is mostly out of our control ( purestake/nimbus and cumulus/aura ), so it's semantically misplaced in **manta**-primitives. I was also thinking some parts of the node don't care about consensus

File: runtime/dolphin/src/nimbus_session_adapter.rs (hunk 8 above)
Comment (Garandor) [reply runtime/dolphin/src/nimbus_session_adapter.rs:35]: in short: yes

purestake does not use the `Session` pallet, we do.
Currently, as nimbus is disabled we need a no-op adapter to Session traits for their types since we don't want to change them directly.
In nimbus stage 1 these will be exchanged to full handlers.

**Comment by Dengjianping** [comment]
Can this new crate be moved to `runtime` or `primitives`?
@ghzlatarev What do you think?

File: node/src/chain_specs/calamari.rs [hunk 10]
```diff
@@ -39,8 +41,13 @@ pub type CalamariChainSpec = sc_service::GenericChainSpec<GenesisConfig, Extensi
...
 ///
 /// The input must be a tuple of individual keys (a single arg for now since we have just one key).
-pub fn calamari_session_keys(keys: AuraId) -> calamari_runtime::opaque::SessionKeys {
-    calamari_runtime::opaque::SessionKeys { aura: keys }
+pub fn calamari_session_keys(keys: AuraId) -> SessionKeys {
+    let unique_dummy_nimbus_id = session_key_primitives::nimbus::from_aura_key(keys.clone());
+    SessionKeys {
+        aura: keys,
+        nimbus: unique_dummy_nimbus_id.clone(),
+        vrf: unique_dummy_nimbus_id.into(),
```
Comment (Dengjianping) [inline node/src/chain_specs/calamari.rs]: In v3.2.0, both new fields will be dummy, but in v3.3.0, will node runners have to update `SessionKeys` by themself?

File: runtime/calamari/Cargo.toml [hunk 11]
```diff
@@ -68,6 +69,9 @@ cumulus-primitives-core = { git = 'https://github.com/paritytech/cumulus.git', d
 cumulus-primitives-timestamp = { git = 'https://github.com/paritytech/cumulus.git', default-features = false, branch = "polkadot-v0.9.18" }
 cumulus-primitives-utility = { git = 'https://github.com/paritytech/cumulus.git', default-features = false, branch = "polkadot-v0.9.18" }
 parachain-info = { git = 'https://github.com/paritytech/cumulus.git', default-features = false, branch = "polkadot-v0.9.18" }
+# Nimbus Dependencies
+nimbus-primitives = { git = "https://github.com/manta-network/nimbus.git", branch = "manta-v0.9.18", default-features = false }
```
Comment (Dengjianping) [inline runtime/calamari/Cargo.toml]: Did you make some changes in this branch? Not sure why you use this branch instead of like `polkadot-v0.9.18`.

File: session-keys/src/lib.rs (hunk 6 above)
Comment (ghzlatarev) [reply session-keys/src/lib.rs]: I'd keep it it in primitives. It just has more parity deps and `nimbus-primitives`, for which `primitives` seems like natural place. Also moonbeam organized it that way.

File: runtime/calamari/Cargo.toml [hunk 12]
```diff
@@ -200,6 +206,9 @@ std = [
   "cumulus-primitives-core/std",
   "cumulus-primitives-timestamp/std",
   "cumulus-primitives-utility/std",
+  "nimbus-primitives/std",
+  "pallet-author-inherent/std",
```
Comment (Dengjianping) [inline runtime/calamari/Cargo.toml:208]: Do we need to benchmark this pallet or be filtered in calamari-runtime?

**Comment by ghzlatarev** [comment]
> Can this new crate be moved to `runtime` or `primitives`? @ghzlatarev What do you think?

Primitives.

File: runtime/calamari/src/lib.rs [hunk 13]
```diff
@@ -90,9 +93,34 @@ pub mod opaque {
...
+            SessionKeys { aura, nimbus, vrf }
+        }
+    }
+
+    pub fn transform_session_keys(_v: AccountId, old: OldSessionKeys) -> SessionKeys {
+        let unique_dummy_nimbus_id =
+            session_key_primitives::nimbus::from_aura_key(old.aura.clone());
+        SessionKeys {
+            aura: old.aura,
+            nimbus: unique_dummy_nimbus_id.clone(),
```
Comment (Dengjianping) [inline runtime/calamari/src/lib.rs]: Seems no need to clones.

File: runtime/calamari/src/lib.rs [hunk 14]
```diff
@@ -752,8 +780,48 @@ pub type Executive = frame_executive::Executive<
...
-    (),
+    UpgradeSessionKeys,
 >;
+// When this is removed, should also remove `OldSessionKeys`.
+pub struct UpgradeSessionKeys;
+impl frame_support::traits::OnRuntimeUpgrade for UpgradeSessionKeys {
+    fn on_runtime_upgrade() -> frame_support::weights::Weight {
+        use opaque::transform_session_keys;
+        // transform_session_keys runs translate() on NextKeys and on QueuedKeys which is (at worst or faster than) 1 read 1 write
+        let validator_set_len: u64 = Session::queued_keys().len().try_into().unwrap();
```
Comment (Dengjianping) [inline runtime/calamari/src/lib.rs:786]: Seems `usize` can `into` `u64`, not 100% sure. Perhaps I'm wrong.

File: session-keys/Cargo.toml [hunk 15]
```diff
@@ -0,0 +1,29 @@
+[package]
+authors = ['Manta Network']
+description = "Primitives for session keys"
+edition = "2021"
+name = "session-key-primitives"
+version = "0.1.0"
```
Comment (Dengjianping) [inline session-keys/Cargo.toml]: license.
And the version is `0.1.0`, kind of misaligned with other crates.

File: session-keys/Cargo.toml [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+manta-primitives = { path = "../primitives", default-features = false }
+nimbus-primitives = { git = "https://github.com/manta-network/nimbus.git", branch = "manta-v0.9.18", default-features = false }
+parity-scale-codec = { version = "3.0.0", default-features = false, features = ["derive"] }
+scale-info = { version = "2.0", default-features = false, features = ["derive"] }
+sp-application-crypto = { git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.18", default-features = false }
+sp-consensus-aura = { git = 'https://github.com/paritytech/substrate.git', default-features = false, branch = "polkadot-v0.9.18" }
+sp-core = { git = 'https://github.com/paritytech/substrate.git', default-features = false, branch = "polkadot-v0.9.18" }
+sp-runtime = { git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.18", default-features = false }
+
+[features]
```
Comment (Dengjianping) [inline session-keys/Cargo.toml]: If nimbus will be benchmarked, I suppose you will add `runtime-benchmarks` in the future.

File: session-keys/src/lib.rs [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+pub mod aura;
+pub mod nimbus;
+pub mod vrf;
+pub use vrf::*;
+
+#[cfg(feature = "std")]
+pub mod helpers;
+
+/// A Trait to lookup keys from AuthorIds
+pub trait KeysLookup<AuthorId, Keys> {
```
Comment (Dengjianping) [inline session-keys/src/lib.rs]: There's a lookup trait in the substrate, no need to define a new one.
https://github.com/paritytech/substrate/blob/master/primitives/runtime/src/traits.rs#L196

If for special reason, never mind.

File: runtime/calamari/Cargo.toml (hunk 11 above)
Comment (Garandor) [reply runtime/calamari/Cargo.toml]: yes. check this commit
https://github.com/Manta-Network/nimbus/commit/009dfaa9f6f7eba47d232d776cb807dc24276d93

It needs to be redone when #571 is in

File: node/src/chain_specs/calamari.rs (hunk 10 above)
Comment (Garandor) [reply node/src/chain_specs/calamari.rs]: Yes. check the PR description

File: runtime/calamari/Cargo.toml (hunk 12 above)
Comment (Garandor) [reply runtime/calamari/Cargo.toml:208]: not in this PR. The pallet is not part of the runtime

File: runtime/calamari/src/lib.rs (hunk 13 above)
Comment (Garandor) [reply runtime/calamari/src/lib.rs]: value gets reused on the next line

File: runtime/calamari/src/lib.rs (hunk 14 above)
Comment (Garandor) [reply runtime/calamari/src/lib.rs:786]: It can't. I tried :D

File: session-keys/Cargo.toml (hunk 16 above)
Comment (Garandor) [reply session-keys/Cargo.toml]: yes, once there is something to benchmark

File: session-keys/Cargo.toml (hunk 15 above)
Comment (Garandor) [reply session-keys/Cargo.toml]: thanks

File: session-keys/src/lib.rs (hunk 17 above)
Comment (Garandor) [reply session-keys/src/lib.rs]: It's a nimbus-specific trait. I added it because they had it.
https://github.com/PureStake/moonbeam/blob/248ae43d6eaecc903fb6f158da1fabbc24a20e5e/pallets/author-mapping/src/lib.rs#L367

Turns out it's only used in `pallet-author-mapping` which we don't need because we use `Session`.
Removing it, thanks for pointing it out.

File: session-keys/src/lib.rs (hunk 6 above)
Comment (Garandor) [reply session-keys/src/lib.rs]: the difference is that they dont have a primitives crate, just a folder
https://github.com/PureStake/moonbeam/tree/master/primitives

i'm ok with doing their thing too, but it'll still be two separate crates.
Also it's really about separation of concerns because manta-specific types and need not be mixed with consensus-specific types

Comment (Garandor) [reply session-keys/src/lib.rs]: @ghzlatarev @Dengjianping moved both primitives crates into a common subfolder as per moonbeam

**Review by ghzlatarev — DISMISSED** [review]

**Comment by Garandor** [comment]
Do not merge before #571

File: CHANGELOG.md [hunk 18]
```diff
@@ -11,10 +11,10 @@
 
 ### Improvements
 - [\#449](https://github.com/Manta-Network/Manta/pull/449) Remove strip from CI, and add strip profile to production.
+- [\#571](https://github.com/Manta-Network/Manta/pull/571) Update upstream dependencies to v0.9.22.
 - [\#563](https://github.com/Manta-Network/Manta/pull/563) Re-implement the `TransactAsset` trait with the unified interface of `FungibleLedger` trait, and `AssetConfig` trait.
 - [\#576](https://github.com/Manta-Network/Manta/pull/576) Unfilter xtokens.transfer_multicurrencies and bump MaxAssetsForTransfer to 2.
 - [\#607](https://github.com/Manta-Network/Manta/pull/607) Turn node client code into library for CLI project
-- [\#609](https://github.com/Manta-Network/Manta/pull/609) Update parameter path from `sdk` to `manta-parameters`.
```
Comment (bhgomes) [inline CHANGELOG.md]: Why was this deleted?

File: pallets/manta-pay/src/rpc.rs [hunk 19]
```diff
@@ -19,25 +19,30 @@
...
 use manta_pay::signer::Checkpoint;
 use sp_api::ProvideRuntimeApi;
 use sp_blockchain::HeaderBackend;
 use sp_runtime::{generic::BlockId, traits::Block};
 
+const SERVER_ERROR: i32 = 1;
+
 /// Pull API
-#[rpc(server)]
+#[rpc(client, server)]
```
Comment (bhgomes) [inline pallets/manta-pay/src/rpc.rs]: Whis is this both a client and server API?

File: primitives/manta/src/assets.rs [hunk 20]
```diff
@@ -352,6 +352,7 @@ where
         asset_id: AssetId,
         account: &C::AccountId,
         amount: Balance,
+        mint: bool,
```
Comment (bhgomes) [inline primitives/manta/src/assets.rs]: Please add to the docstring what `mint` does and rename it to a boolean name like `should_mint` or `is_mint` whatever the appropriate thing is.

File: primitives/session-keys/src/lib.rs [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+// You should have received a copy of the GNU General Public License
+// along with Manta.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Primitives for session keys
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub mod aura;
+pub mod nimbus;
+pub mod vrf;
+pub use vrf::*;
```
Comment (bhgomes) [inline primitives/session-keys/src/lib.rs]: Why re-export `vrf` keys and not the others?

File: primitives/session-keys/src/nimbus.rs [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+use nimbus_primitives::NimbusId;
+use sp_application_crypto::{sr25519, UncheckedFrom};
+
+/// Clones an aura pubkey and represents it as a nimbus pubkey
+/// Note: Not implementing From/Into because neither Nimbus nor Aura are our crates
+pub fn from_aura_key(aura_id: AuraId) -> NimbusId {
+    let aura_as_sr25519: sr25519::Public = aura_id.into();
+    let sr25519_as_bytes: [u8; 32] = aura_as_sr25519.into();
+    sr25519::Public::unchecked_from(sr25519_as_bytes).into()
+}
```
Comment (bhgomes) [inline primitives/session-keys/src/nimbus.rs]: This is not cryptographically sound. We are representing a key with one domain under another domain. If this is needed for dummy keys, then it should be very clear in the function name and documentation that this is not meant to be used in a safe environment.

File: primitives/session-keys/src/vrf.rs [hunk 23]
```diff
@@ -0,0 +1,71 @@
...
+impl From<NimbusId> for VrfId {
+    /// Reinterprets Nimbus as VRFId.
+    /// No corresponding private key to that VRFId will exist,
+    /// **use this only for dummy public keys**
+    fn from(nimbus_id: NimbusId) -> VrfId {
+        let nimbus_as_sr25519: sr25519::Public = nimbus_id.into();
+        let sr25519_as_bytes: [u8; 32] = nimbus_as_sr25519.into();
+        sr25519::Public::unchecked_from(sr25519_as_bytes).into()
+    }
+}
```
Comment (bhgomes) [inline primitives/session-keys/src/vrf.rs]: Don't use a `From` implementation for this if it's unsound, add an explicit function.

File: primitives/session-keys/src/vrf.rs [hunk 24]
```diff
@@ -0,0 +1,71 @@
...
+
+/// The ConsensusEngineId for VRF keys
+pub const VRF_ENGINE_ID: ConsensusEngineId = *b"rand";
```
Comment (bhgomes) [inline primitives/session-keys/src/vrf.rs:38]: Is this the domain tag other people are using?

File: primitives/session-keys/src/vrf.rs [hunk 25]
```diff
@@ -0,0 +1,71 @@
...
+
+#[test]
+fn nimbus_to_vrf_id() {
+    for x in 0u8..10u8 {
+        let nimbus_id: NimbusId = sr25519::Public::unchecked_from([x; 32]).into();
+        let expected_vrf_id: VrfId = sr25519::Public::unchecked_from([x; 32]).into();
+        let nimbus_to_vrf_id: VrfId = nimbus_id.into();
+        assert_eq!(expected_vrf_id, nimbus_to_vrf_id);
+    }
+}
```
Comment (bhgomes) [inline primitives/session-keys/src/vrf.rs]: What is this testing?

File: runtime/dolphin/src/nimbus_session_adapter.rs [hunk 26]
```diff
@@ -0,0 +1,81 @@
...
+    {
+    }
+
+    fn on_disabled(_: u32) {}
+
+    fn on_before_session_ending() {}
+}
+
+use crate::AccountId;
+use session_key_primitives::vrf::VrfSessionKey;
```
Comment (bhgomes) [inline runtime/dolphin/src/nimbus_session_adapter.rs]: Move up to the top.

File: runtime/manta/src/lib.rs [hunk 27]
```diff
@@ -103,6 +103,13 @@ pub mod opaque {
             pub aura: Aura,
         }
     }
+    impl SessionKeys {
+        /// generates SessionKey, discarding key types not used on manta
+        pub fn new<T, U>(tuple: (AuraId, T, U)) -> SessionKeys {
+            let (aura, _, _) = tuple;
+            SessionKeys { aura }
+        }
```
Comment (bhgomes) [inline runtime/manta/src/lib.rs]:
```suggestion
        /// Generates a [`SessionKey`], discarding key types that are not currently used.
        #[inline]
        pub fn new<T, U>((aura, _, _): (AuraId, T, U)) -> Self {
            Self { aura }
        }
```

**Review by bhgomes — DISMISSED** [review]
Does this PR depend on #571 being merged first? If so, please leave a note in the description and mark it with a DO-NOT-MERGE label. Also, this PR does not actually close #486, so just mark it as related but not closing.

File: pallets/manta-pay/src/rpc.rs (hunk 19 above)
Comment (Garandor) [reply pallets/manta-pay/src/rpc.rs]: @Dengjianping please address this on #571 if it doesn't explicitly need to be a client

File: primitives/manta/src/assets.rs (hunk 20 above)
Comment (Garandor) [reply primitives/manta/src/assets.rs]: This code is outdated, it has been addressed in
https://github.com/Manta-Network/Manta/pull/571/files#diff-e7fdf4a079f6a2130e9cc7ef71f550c7ab4bdfe6f3b85a6a0f1ad8a1863091ceR327

File: CHANGELOG.md (hunk 18 above)
Comment (Garandor) [reply CHANGELOG.md]: it was mistakenly added in a previous PR

File: primitives/session-keys/src/vrf.rs (hunk 23 above)
Comment (Garandor) [reply primitives/session-keys/src/vrf.rs]: this crate was forked from https://github.com/PureStake/moonbeam/pull/1480 and that is what they use

These Vrf keys are placeholders only and the whole functionality will move to a nnew `vrf-consensus` crate when it's built, at which point this crutch will be removed.
I'll make it more explicit

File: primitives/session-keys/src/vrf.rs (hunk 24 above)
Comment (Garandor) [reply primitives/session-keys/src/vrf.rs:38]: yes. this crate is forked from https://github.com/PureStake/moonbeam/pull/1480

File: primitives/session-keys/src/vrf.rs (hunk 25 above)
Comment (Garandor) [reply primitives/session-keys/src/vrf.rs]: identity of directly created and converted-from-nimbus public vrf keys ( i did not write this code )

File: primitives/session-keys/src/nimbus.rs (hunk 22 above)
Comment (Garandor) [reply primitives/session-keys/src/nimbus.rs]: thanks, added

File: primitives/session-keys/src/lib.rs (hunk 21 above)
Comment (Garandor) [reply primitives/session-keys/src/lib.rs]: fixed to reexport all 3 pubkey types

File: primitives/session-keys/src/vrf.rs (hunk 23 above)
Comment (Garandor) [reply primitives/session-keys/src/vrf.rs]: done

File: runtime/dolphin/src/nimbus_session_adapter.rs (hunk 26 above)
Comment (Garandor) [reply runtime/dolphin/src/nimbus_session_adapter.rs]: done

File: runtime/manta/src/lib.rs (hunk 27 above)
Comment (Garandor) [reply runtime/manta/src/lib.rs]: so much better, thanks!

**Comment by ghzlatarev** [comment]
> Does this PR depend on #571 being merged first? If so, please leave a note in the description and mark it with a DO-NOT-MERGE label. Also, this PR does not actually close #486, so just mark it as related but not closing.

If we don't use `closes` it won't show up in the linked PR section in issues, so that's why we've been closing and reopening them.

**Review by ghzlatarev — DISMISSED** [review]

**Review by ghzlatarev — DISMISSED** [review]

**Review by ghzlatarev — DISMISSED** [review]

File: primitives/session-keys/Cargo.toml [hunk 28]
```diff
@@ -0,0 +1,30 @@
+[package]
+authors = ['Manta Network']
+description = "Primitives for session keys"
+edition = "2021"
+license = 'GPL-3.0'
+name = "session-key-primitives"
+version = "1.0.0"
```
Comment (Dengjianping) [inline primitives/session-keys/Cargo.toml]: If the version of crate doesn't align to others, what will the next version be?
@ghzlatarev Do you think we should align the version with others'?

File: primitives/session-keys/Cargo.toml [hunk 29]
```diff
@@ -0,0 +1,30 @@
...
+[dependencies]
+manta-primitives = { path = "../manta", default-features = false }
+nimbus-primitives = { git = "https://github.com/manta-network/nimbus.git", branch = "polkadot-v0.9.22", default-features = false }
+parity-scale-codec = { version = "3.0.0", default-features = false, features = ["derive"] }
+scale-info = { version = "2.0", default-features = false, features = ["derive"] }
+sp-application-crypto = { git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.22", default-features = false }
+sp-consensus-aura = { git = 'https://github.com/paritytech/substrate.git', default-features = false, branch = "polkadot-v0.9.22" }
+sp-core = { git = 'https://github.com/paritytech/substrate.git', default-features = false, branch = "polkadot-v0.9.22" }
+sp-runtime = { git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.22", default-features = false }
+
```
Comment (Dengjianping) [inline primitives/session-keys/Cargo.toml:18]: Maybe you should add `runtime-benchmarks` feature in the future. I suppose it should be enabled if we benchmark consensus related pallets.

File: primitives/session-keys/src/vrf.rs [hunk 30]
```diff
@@ -0,0 +1,68 @@
...
+pub type VrfSignature = vrf_crypto::Signature;
+
+sp_application_crypto::with_pair! {
+    /// A vrf key pair
+    pub type VrfPair = vrf_crypto::Pair;
+}
+
+#[test]
+fn converted_dummy_id_is_same_as_newly_created_id() {
+    for x in 0u8..10u8 {
```
Comment (Dengjianping) [inline primitives/session-keys/src/vrf.rs:62]: why `0..10`?

File: primitives/session-keys/Cargo.toml [hunk 31]
```diff
@@ -0,0 +1,30 @@
...
+description = "Primitives for session keys"
+edition = "2021"
+license = 'GPL-3.0'
+name = "session-key-primitives"
+version = "1.0.0"
+
+[dependencies]
+manta-primitives = { path = "../manta", default-features = false }
+nimbus-primitives = { git = "https://github.com/manta-network/nimbus.git", branch = "polkadot-v0.9.22", default-features = false }
+parity-scale-codec = { version = "3.0.0", default-features = false, features = ["derive"] }
```
Comment (Dengjianping) [inline primitives/session-keys/Cargo.toml]: `codec` and `scale-info` should be bumped, please take a look at other crates.
It should be `2.1.2` and `3.1.2`.

File: primitives/session-keys/Cargo.toml [hunk 32]
```diff
@@ -0,0 +1,30 @@
...
+scale-info = { version = "2.0", default-features = false, features = ["derive"] }
+sp-application-crypto = { git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.22", default-features = false }
```
Comment (Dengjianping) [inline primitives/session-keys/Cargo.toml]: `substrate` => `substrate.git`

File: primitives/session-keys/Cargo.toml [hunk 33]
```diff
@@ -0,0 +1,30 @@
...
+sp-consensus-aura = { git = 'https://github.com/paritytech/substrate.git', default-features = false, branch = "polkadot-v0.9.22" }
+sp-core = { git = 'https://github.com/paritytech/substrate.git', default-features = false, branch = "polkadot-v0.9.22" }
+sp-runtime = { git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.22", default-features = false }
```
Comment (Dengjianping) [inline primitives/session-keys/Cargo.toml]: `substrate` => `substrate.git`

File: runtime/calamari/src/lib.rs [hunk 34]
```diff
@@ -750,8 +775,48 @@ pub type Executive = frame_executive::Executive<
...
```
Comment (Dengjianping) [inline runtime/calamari/src/lib.rs:786]: I suppose it will never panic our runtime, right?

File: primitives/session-keys/Cargo.toml (hunk 28 above)
Comment (ghzlatarev) [reply primitives/session-keys/Cargo.toml]: Yeah I forgot to change this earlier.

File: runtime/calamari/src/lib.rs (hunk 34 above)
Comment (Garandor) [reply runtime/calamari/src/lib.rs:786]: the session keys DB on KMA is <30 entries
and the upgrade reserves half a block for good measure ( if it takes longer, it uses the actual weight )
30 db reads/writes should not have a chance to overflow block execution limits

File: primitives/session-keys/src/vrf.rs (hunk 30 above)
Comment (Garandor) [reply primitives/session-keys/src/vrf.rs:62]: it's a simple sanity check that the conversion is performing as expected.
It's not meant to be exhaustive

File: primitives/session-keys/Cargo.toml (hunk 29 above)
Comment (Garandor) [reply primitives/session-keys/Cargo.toml:18]: It will be enabled in https://github.com/Manta-Network/Manta/pull/582

File: primitives/session-keys/Cargo.toml (hunk 33 above)
Comment (Garandor) [reply primitives/session-keys/Cargo.toml]: Addressed, thx

**Review by ghzlatarev — APPROVED** [review]

**Review by ghzlatarev — APPROVED** [review]

**Review by bhgomes — APPROVED** [review]
